### PR TITLE
Move RTPSDomain mutex lock to callers of find_local_participant [10843]

### PR DESCRIPTION
**Pull request by MiguelCompany**
This prevents the race detected on #1819 by ensuring that the `RTPSDomain` mutex is taken when a new entry is added to `RTPSParticipant::m_allReaderList` or `RTPSParticipant::m_allWriterList`

### CODE REVIEW

**Review by fujitatomoya — COMMENTED** [review]
as far as i see, this makes sense to me.

**Review by IkerLuengo — DISMISSED** [review]
:+1:

File: src/cpp/rtps/RTPSDomain.cpp [hunk 1]
```diff
@@ -434,12 +434,13 @@ RTPSParticipantImpl* RTPSDomainImpl::find_local_participant(
...
         const GUID_t& reader_guid)
 {
-    std::lock_guard<std::mutex> guard(RTPSDomain::m_mutex);
+    std::unique_lock<std::mutex> lock(RTPSDomain::m_mutex);
     for (const RTPSDomain::t_p_RTPSParticipant& participant : RTPSDomain::m_RTPSParticipants)
     {
         if (participant.second->getGuid().guidPrefix == reader_guid.guidPrefix)
         {
             // Participant found, forward the query
+            lock.unlock();
```
Comment (VladimirVassiliev) [inline src/cpp/rtps/RTPSDomain.cpp:443]: This lock cannot be released here, as it is protecting m_RTPSParticipants and the participant reference.

File: src/cpp/rtps/RTPSDomain.cpp [hunk 2]
```diff
@@ -450,12 +451,13 @@ RTPSReader* RTPSDomainImpl::find_local_reader(
...
         const GUID_t& writer_guid)
 {
-    std::lock_guard<std::mutex> guard(RTPSDomain::m_mutex);
+    std::unique_lock<std::mutex> lock(RTPSDomain::m_mutex);
     for (const RTPSDomain::t_p_RTPSParticipant& participant : RTPSDomain::m_RTPSParticipants)
     {
         if (participant.second->getGuid().guidPrefix == writer_guid.guidPrefix)
         {
             // Participant found, forward the query
+            lock.unlock();
```
Comment (JLBuenoLopez-eProsima) [inline src/cpp/rtps/RTPSDomain.cpp:460]: The same issue also applies here.

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]
There are several build warnings. Some of them I think are due to not being rebased with the latest changes in `master` (for `2.5.0` the test types were updated). However, some of the warnings seems to be introduced by this PR. I suggest rebasing to `master`.

**Comment by MiguelCompany** [comment]
@JLBuenoLopez-eProsima This PR is not correct. What we should do is have another mutex on RTPSDomainParticipantImpl protecting the endpoint collections.

That would be easy to do. The difficult part comes in designing a regression test.

**Comment by JLBuenoLopez-eProsima** [comment]
Close in favor #2506

**Pull request closed without merging**